src/backend/distributed/operations/repair_shards.c

```diff
@@ -158,6 +176,111 @@ master_copy_shard_placement(PG_FUNCTION_ARGS)
 }
 
 
+/*
+ * ColocationSize returns the size in bytes of a set of colocated tables.
+ */
+int64
+ColocationSize(List *colocatedShardList, char *workerNodeName, uint32 workerNodePort)
+{
+	uint32 connectionFlag = 0;
+
+	/* we skip child tables of a partitioned table if this boolean variable is true */
+	bool optimizePartitionCalculations = true;
+	StringInfo tableSizeQuery = GenerateSizeQueryOnMultiplePlacements(colocatedShardList,
+																	  TOTAL_RELATION_SIZE,
+																	  optimizePartitionCalculations);
+
+	MultiConnection *connection = GetNodeConnection(connectionFlag, workerNodeName,
+													workerNodePort);
+	PGresult *result = NULL;
+	int queryResult = ExecuteOptionalRemoteCommand(connection, tableSizeQuery->data,
+												   &result);
+
+	if (queryResult != RESPONSE_OKAY)
+	{
+		ereport(ERROR, (errcode(ERRCODE_CONNECTION_FAILURE),
+						errmsg("cannot get the size because of a connection error")));
+	}
+
+	List *sizeList = ReadFirstColumnAsText(result);
+	if (list_length(sizeList) != 1)
+	{
+		ereport(ERROR, (errmsg(
+							"received wrong number of rows from worker, expected 1 received %d",
+							list_length(sizeList))));
+	}
+
+	StringInfo tableSizeStringInfo = (StringInfo) linitial(sizeList);
+	char *tableSizeString = tableSizeStringInfo->data;
+	uint64 tableSize = SafeStringToUint64(tableSizeString);
+
+	bool raiseErrors = true;
+	PQclear(result);
+	ClearResults(connection, raiseErrors);
+
+	return tableSize;
+}
+
+
+/*
+ * CheckSpaceConstraints checks there is enough space to place the colocation
+ * on the node that the connection is connected to.
+ */
+static
+DeferredErrorMessage *
+CheckSpaceConstraints(MultiConnection *connection, int64 colocationSize)
+{
+	if (!CheckAvailableSpace)
+	{
+		return NULL;
+	}
+	int64 diskAvailable = 0;
+	int64 diskSize = 0;
+	StringInfo errorDetail = makeStringInfo();
+	if (ForceDiskAvailable != -1)
+	{
+		diskAvailable = ForceDiskAvailable;
+	}
+	else
+	{
+		diskAvailable = ExecuteRemoteInt64Command(connection,
+												  "select pg_catalog.citus_disk_available()");
+	}
+	if (ForceDiskSize != -1)
+	{
+		diskSize = ForceDiskSize;
+	}
+	else
+	{
+		diskSize = ExecuteRemoteInt64Command(connection,
+											 "select pg_catalog.citus_disk_size()");
+	}
+
+	if (diskAvailable < colocationSize)
+	{
+		appendStringInfo(errorDetail, "actual space %ld, required space %ld",
```

Comment:
I am adding the units here "bytes" in the error message so that it is clear how much space there is.

Reply:
Also in case the actual space is "forced", it seems better to say this in the error message for easier identification of problems